sources/web/datalab/fileSearch.ts

```diff
@@ -0,0 +1,143 @@
+const clientResultSize = 20; 
+
+/**
+ * Implements the file search/filter request handling.
+ * @param request the incoming search request.
+ * @param response the outgoing search response.
+ */
+function requestHandler(request: http.ServerRequest, response: http.ServerResponse): void {
+  const parsedUrl = url.parse(request.url, true);
+  const pattern = parsedUrl.query['pattern'];  
+  const statusCheck = parsedUrl.query['status'];
+
+  response.writeHead(200, { 'Content-Type': 'application/json' });
+  if (pattern !== undefined) {
+    const results = filter(pattern);
+    response.write(JSON.stringify(results));
+  } else if (statusCheck !== undefined) {
+    response.write(JSON.stringify({
+      tooManyFiles: tooManyFiles
+    }));
+  }
+  response.end();
+}
+
+/**
+ * Builds an index of all files in the content directory to make search faster
+ */
+export function indexFiles(): void {
+  const startTime = process.hrtime();
+  index(appSettings.contentDir + '/', () => {
+    const indexTime = process.hrtime(startTime);
+    if (tooManyFiles) {
+      logging.getLogger().error('Hit file index maximum limit, searching files should be disabled');
+    } else {
+      logging.getLogger().info('Finished indexing ' + fileIndex.length + ' files in ' + indexTime[0] + ' seconds');
+    }
+  });
+}
+
+/**
+ * Recursively indexes all files under the given search path into the global fileIndex object
+ * @param searchpath the path prefix to start indexing
+ */
+function index(searchpath: string, callback: Function): void {
+  // stop the search if we've reached the file count limit
+  if (fileIndex.length >= fileCountLimit) {
+    tooManyFiles = true;
+    return callback();
+  }
+  fs.readdir(searchpath, (err, list) => {
+    if (err) {
+      logging.getLogger().error('Could not read dir ' + searchpath);
+      return callback();
+    }
+    let remaining = list.length;
+    if (remaining === 0) {
+      return callback();
+    }
+    list.forEach((file) => {
+      // ignore hidden files/dirs
+      if (file[0] === '.') {
+        if (--remaining === 0) {
+          callback();
+        }
+        return;
+      }
+
+      const filename = path.join(searchpath, file);
+      fs.lstat(filename, (err, stat) => {
+        if (err) {
+          remaining--;
+          callback();
+        } else {
+          if (stat.isDirectory()) {
+            index(filename, () => {
+              if (--remaining === 0) {
+                callback();
+              }
+            });
+          } else {
+            fileIndex.push(filename.substr(appSettings.contentDir.length + 1));
+            if (--remaining === 0) {
+              callback();
+            }
+          }
+        }
+      });
+    });
+  });
+}
+
+/**
+ * Filters the file index based on the provided pattern
+ * @param pattern the search pattern
+ * @returns a list of matches that are superstrings of pattern
+ */
+function filter(pattern: string): string[] {
```

Comment:
I don't see an issue with the current naming or implementation. It's scoped to the fileSearch file, so the simple `filter` name should do.
Turning it to a pure function is a good idea though.